### ISSUE TITLE
[BUG] DHCP renewal and rebinding fail constantly in kube-vip

### ISSUE THREAD

**Issue by starbops** [harvester/harvester#6637]
**Describe the bug**

TL;DR: DHCP lease renewal and rebinding operations in kube-vip fail, resulting in a potential loss/conflict of IP addresses for the load balancer service. The DHCP server issues NAK responses, indicating the lease has not been found. This issue allegedly has been fixed upstream in the kube-vip project. We need to update our dependency and verify with our use cases to see whether the issue has been fixed.

Below are our findings:

## Kube-vip

Creating the load balancer service object:

```
time="2024-09-23T15:57:34Z" level=info msg="Creating new macvlan interface for DHCP [vip-fdd3b9c3]"
time="2024-09-23T15:57:34Z" level=info msg="Generated mac: 00:00:6C:d9:71:95"
time="2024-09-23T15:57:34Z" level=info msg="New interface [vip-fdd3b9c3] mac is 00:00:6c:d9:71:95"
time="2024-09-23T15:57:37Z" level=info msg="DHCP VIP [0.0.0.0] for [default/blog-loadbalancer] "
time="2024-09-23T15:57:37Z" level=info msg="[service] adding VIP [0.0.0.0] for [default/blog-loadbalancer]"
time="2024-09-23T15:57:37Z" level=info msg="[service] synchronised in 3747ms"
```

When kube-vip tried to renew the DHCP lease for the load balancer service IP address:

```
time="2024-09-23T16:28:12Z" level=error msg="renew failed, error: got an error while processing the request: no matching response packet received"
```

When kube-vip tried to rebind the DHCP lease for the load balancer service IP address:

```
time="2024-09-23T16:50:42Z" level=error msg="rebind failed, error: got an error while processing the request: no matching response packet received"
```

Deleting the load balancer service object:

```
time="2024-09-24T01:21:46Z" level=info msg="[LOADBALANCER] Stopping load balancers"
time="2024-09-24T01:21:46Z" level=info msg="[VIP] Releasing the Virtual IP [172.19.32.94]"
time="2024-09-24T01:21:46Z" level=info msg="release, lease: &{Offer:DHCPv4(xid=0x45696ed9 hwaddr=00:00:6c:d9:71:95 msg_type=OFFER, your_ip=172.19.32.94, server_ip=172.19.32.1) ACK:DHCPv4(xid=0x45696ed9 hwaddr=00:00:6c:d9:71:95 msg_type=ACK, your_ip=172.19.32.94, server_ip=172.19.32.1) CreationTime:2024-09-23 15:57:37.694602994 +0000 UTC m=+383.996501377}"
time="2024-09-24T01:21:47Z" level=info msg="Removed [fdd3b9c3-f7ea-4f82-a645-b3844b0d75a4] from manager, [0] advertised services remain"
time="2024-09-24T01:21:47Z" level=info msg="service [default/blog-loadbalancer] has been deleted"
```

## Dnsmasq

(Given that the lease expiration time is set to one hour)

The DORA process found on the DHCP server side during the load balancer creation:

```
Sep 23 15:57:34 hp-171 dnsmasq-dhcp[1370339]: 1164537561 available DHCP range: 172.19.32.51 -- 172.19.32.100
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 DHCPDISCOVER(br-data) 00:00:6c:d9:71:95
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 tags: br-data
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 DHCPOFFER(br-data) 172.19.32.94 00:00:6c:d9:71:95
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: Ignoring duplicate dhcp-option 6
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: Ignoring duplicate dhcp-option 3
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: Ignoring duplicate dhcp-option 1
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 requested options: 1:netmask, 3:router, 15:domain-name, 6:dns-server
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 next server: 172.19.32.1
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  1 option: 53 message-type  2
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option: 54 server-identifier  172.19.32.1
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option: 51 lease-time  1h
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option: 58 T1  30m
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option: 59 T2  52m30s
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option: 28 broadcast  172.19.32.255
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option:  6 dns-server  1.1.1.1
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option:  3 router  172.19.31.1
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option:  1 netmask  255.255.255.0
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 available DHCP range: 172.19.32.51 -- 172.19.32.100
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 DHCPREQUEST(br-data) 172.19.32.94 00:00:6c:d9:71:95
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 tags: br-data
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 DHCPACK(br-data) 172.19.32.94 00:00:6c:d9:71:95
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: Ignoring duplicate dhcp-option 6
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: Ignoring duplicate dhcp-option 3
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: Ignoring duplicate dhcp-option 1
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 requested options: 1:netmask, 3:router, 15:domain-name, 6:dns-server
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 next server: 172.19.32.1
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  1 option: 53 message-type  5
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option: 54 server-identifier  172.19.32.1
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option: 51 lease-time  1h
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option: 58 T1  30m
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option: 59 T2  52m30s
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option: 28 broadcast  172.19.32.255
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option:  6 dns-server  1.1.1.1
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option:  3 router  172.19.31.1
Sep 23 15:57:37 hp-171 dnsmasq-dhcp[1370339]: 1164537561 sent size:  4 option:  1 netmask  255.255.255.0
```

After 30 minutes, the DHCP renewal was rejected:

```
Sep 23 16:27:37 hp-171 dnsmasq-dhcp[1370339]: 418945112 available DHCP range: 172.19.32.51 -- 172.19.32.100
Sep 23 16:27:37 hp-171 dnsmasq-dhcp[1370339]: 418945112 DHCPREQUEST(br-data) 0.0.0.0 00:00:6c:d9:71:95
Sep 23 16:27:37 hp-171 dnsmasq-dhcp[1370339]: 418945112 DHCPNAK(br-data) 0.0.0.0 00:00:6c:d9:71:95 lease not found
Sep 23 16:27:37 hp-171 dnsmasq-dhcp[1370339]: 418945112 broadcast response
Sep 23 16:27:37 hp-171 dnsmasq-dhcp[1370339]: 418945112 sent size:  1 option: 53 message-type  6
Sep 23 16:27:37 hp-171 dnsmasq-dhcp[1370339]: 418945112 sent size:  4 option: 54 server-identifier  172.19.32.1
Sep 23 16:27:37 hp-171 dnsmasq-dhcp[1370339]: 418945112 sent size: 15 option: 56 message  6c:65:61:73:65:20:6e:6f:74:20:66:6f:75:6e...
Sep 23 16:27:42 hp-171 dnsmasq-dhcp[1370339]: 418945112 available DHCP range: 172.19.32.51 -- 172.19.32.100
Sep 23 16:27:42 hp-171 dnsmasq-dhcp[1370339]: 418945112 DHCPREQUEST(br-data) 0.0.0.0 00:00:6c:d9:71:95
Sep 23 16:27:42 hp-171 dnsmasq-dhcp[1370339]: 418945112 DHCPNAK(br-data) 0.0.0.0 00:00:6c:d9:71:95 lease not found
Sep 23 16:27:42 hp-171 dnsmasq-dhcp[1370339]: 418945112 broadcast response
Sep 23 16:27:42 hp-171 dnsmasq-dhcp[1370339]: 418945112 sent size:  1 option: 53 message-type  6
Sep 23 16:27:42 hp-171 dnsmasq-dhcp[1370339]: 418945112 sent size:  4 option: 54 server-identifier  172.19.32.1
Sep 23 16:27:42 hp-171 dnsmasq-dhcp[1370339]: 418945112 sent size: 15 option: 56 message  6c:65:61:73:65:20:6e:6f:74:20:66:6f:75:6e...
Sep 23 16:27:52 hp-171 dnsmasq-dhcp[1370339]: 418945112 available DHCP range: 172.19.32.51 -- 172.19.32.100
Sep 23 16:27:52 hp-171 dnsmasq-dhcp[1370339]: 418945112 DHCPREQUEST(br-data) 0.0.0.0 00:00:6c:d9:71:95
Sep 23 16:27:52 hp-171 dnsmasq-dhcp[1370339]: 418945112 DHCPNAK(br-data) 0.0.0.0 00:00:6c:d9:71:95 lease not found
Sep 23 16:27:52 hp-171 dnsmasq-dhcp[1370339]: 418945112 broadcast response
Sep 23 16:27:52 hp-171 dnsmasq-dhcp[1370339]: 418945112 sent size:  1 option: 53 message-type  6
Sep 23 16:27:52 hp-171 dnsmasq-dhcp[1370339]: 418945112 sent size:  4 option: 54 server-identifier  172.19.32.1
Sep 23 16:27:52 hp-171 dnsmasq-dhcp[1370339]: 418945112 sent size: 15 option: 56 message  6c:65:61:73:65:20:6e:6f:74:20:66:6f:75:6e...
```

After 22.5 minutes, the DHCP rebinding was not successful either (no meaningful messages were found on the server side):

```
Sep 23 16:50:07 hp-171 dnsmasq-dhcp[1370339]: 566851277 available DHCP range: 172.19.32.51 -- 172.19.32.100
Sep 23 16:50:12 hp-171 dnsmasq-dhcp[1370339]: 566851277 available DHCP range: 172.19.32.51 -- 172.19.32.100
Sep 23 16:50:22 hp-171 dnsmasq-dhcp[1370339]: 566851277 available DHCP range: 172.19.32.51 -- 172.19.32.100
```

The DHCP server cannot find the corresponding lease when receiving the DHCP release packet:

```
Sep 24 01:21:46 hp-171 dnsmasq-dhcp[1370339]: 2363481030 available DHCP range: 172.19.32.51 -- 172.19.32.100
Sep 24 01:21:46 hp-171 dnsmasq-dhcp[1370339]: 2363481030 DHCPRELEASE(br-data) 172.19.32.94 00:00:6c:d9:71:95 unknown lease
```

This is mostly due to the inconsistency of the interface used for the DHCP DORA and renewal/rebinding process.

Upgrading to kube-vip v0.8.1 will hit harvester/harvester#6563. The renewal still has an issue, but the rebinding works, which kind of relieve the issue we raised here.

**To Reproduce**

For the VM workload load balancer scenario (Harvester cluster):
1. Prepare a external DHCP server with an appropriate configuration (for easy reproduction it is recommended to set the lease expiration to a smaller value)
2. Create a VM
3. Create a load balancer with DHCP IPAM selected
4. Wait for the lease to expire after the load balancer acquire an IP address from the DHCP server
5. Check the logs from the kube-vip pod

For the LB service scenario (guest Kubernetes cluster):
1. Prepare a external DHCP server with an appropriate configuration (for easy reproduction it is recommended to set the lease expiration to a smaller value)
2. Create a guest cluster from the Rancher dashboard (enable Harvester cloud provider)
3. After the guest cluster is created, create a deployment (nginx) with a LB service (DHCP IPAM)
4. Wait for the lease to expire after the LB service acquire an IP address from the DHCP server
5. Check the logs from the kube-vip pod (the one in the guest cluster)

**Expected behavior**

DHCP renewal should work for both LB services in guest clusters and VM workload LBs in Harvester clusters.

**Support bundle**

N/A

**Environment**
 - Harvester
   - ISO version: v1.3.1
   - harvester-load-balancer version: v0.3.0
   - kube-vip version: v0.6.0
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): KVM
 - Guest cluster
   - RKE2 version: v1.28.13
   - harvester-cloud-provider version: v0.2.1
   - kube-vip version: v0.6.0

**Additional context**

This is a tracking issue from our side for its upstream counterpart. A deep analysis can be found at kube-vip/kube-vip#871.

**Comment by starbops** [comment]
After testing, I can confirm that both kube-vip v0.8.2 and v0.8.3 still hit #6563, i.e., the renewal failed, but rebinding succeeded.

**Comment by bpedersen2** [comment]
This looks like kube-vip v0.9.1 fixed that (https://github.com/kube-vip/kube-vip/pull/897).
